### ISSUE TITLE
[BUG] Typescript error when using useTransform to combine 2 useTransforms

### ISSUE THREAD

**Issue by tarngerine** [framer/motion#840]
**2. Describe the bug**
I have 2 useTransforms (x, y) which I want to combine into a string to use as a CSS transform (not sure of what other way to do this — top/left isn't performant)
<img width="684" alt="image" src="https://user-images.githubusercontent.com/1629901/98420642-4230d800-204d-11eb-8999-948debf5736d.png">

```
  const progress = useSpring(0, {spring});
  const x = useTransform(progress, [0, 1], [0, 100]);
  const y = useTransform(progress, [0, 1], [0, 100]);
  const t = useTransform([x, y], ([x:, y]) => `translate(${x}px, ${y}px)`);
```

But when I pass them into a 3rd useTransform to combine, typescript complains

```
TypeScript error in /Users/tarng/Code/playspace/src/av/Stream.tsx(21,26):
No overload matches this call.
  Overload 1 of 3, '(input: MotionValue<unknown>, transformer: SingleTransformer<unknown, unknown>): MotionValue<unknown>', gave the following error.
    Argument of type 'MotionValue<number>[]' is not assignable to parameter of type 'MotionValue<unknown>'.
      Type 'MotionValue<number>[]' is missing the following properties from type 'MotionValue<unknown>': current, prev, timeDelta, lastUpdated, and 19 more.  TS2769

    19 |   const x = useTransform(progress, [0, 1], [0, 100]);
    20 |   const y = useTransform(progress, [0, 1], [0, 100]);
  > 21 |   const t = useTransform([x, y], ([x, y]) => `translate(${x}px, ${y}px)`);
       |                          ^
```

if i try this on codepen without strict typescript, it throws this error:

<img width="543" alt="image" src="https://user-images.githubusercontent.com/1629901/98420614-30e7cb80-204d-11eb-9da6-59672d509688.png">

**3. IMPORTANT: Provide a CodeSandbox reproduction of the bug**
https://codesandbox.io/s/framer-motion-2-animatesharedlayout-animate-between-different-components-forked-qmgro?file=/src/App.js

**4. Steps to reproduce**

Load the prototype, see typescript or type error

**5. Expected behavior**

It should run

**6. Video or screenshots**
N/A
**7. Environment details**
Chrome, macos

**Comment by tarngerine** [comment]
complains even with the sample code from the overload definition
<img width="1459" alt="image" src="https://user-images.githubusercontent.com/1629901/98420844-aa7fb980-204d-11eb-960d-cb0a63a48a7e.png">

**Comment by huntercaron** [comment]
Ive also run into this, not sure how to fix though

**Comment by tarngerine** [comment]
i cry

```

  const x = useTransform([progress, mousePos.x] as any, ((progress: any, start: any) => lerp(start, targetX, progress)) as any);
  const y = useTransform([progress, mousePos.y] as any, ((progress: any, start: any) => lerp(start, targetY, progress)) as any);
```

**Comment by huntercaron** [comment]
@tarngerine  yeah I just poked around a bit I think thats as good as your gonna get for now. I'll see if I can try to understand the TS in there well enough to fix it but I'm not sure

**Comment by kena0ki** [comment]
A type of parameter of useTransform is defined as `input: MotionValue<string | number>[]`.
https://github.com/framer/motion/blob/dc571a6625dafa1d195d569db6b84c449fcc4140/src/value/use-transform.ts#L163
When strictFunctionTypes in tsconfig is set to true, MotionValue<number> is not assignable to MotionValue<string | number>.
This is because MotionValue has functions which have parameter of type string | number, and `number => void` is not assignable to `(number|string) => void`.
You can see the error in the link below and if you set strictFunctionTypes to false, the error will disappear.
https://www.typescriptlang.org/play?#code/PTAEFcDsEsHtNAFwJ4AcCmpUEMBO2BbdRdXAWACgAbYpAQVAF5QAKADwC5RJwCAjUgEomAPlABvAL4BuSjURIAQk1aduvAblAAfUAGdEuaJADmwxmKmyKiBs0SLrDlbelA

Probably useTransform needs conditional types.

**Comment by kena0ki** [comment]
The version of framer-motion in the above codesandbox is too old.

Minimal repro:
https://codesandbox.io/s/framer-motion-simple-animation-forked-8p25e?file=/src/index.tsx

**Comment by goszczynskip** [comment]
If you want a temp fix you can grab my types augmentation. Just make sure it's discoverable for tsc. 

```typescript
import { MotionValue } from 'framer-motion'

declare module 'framer-motion' {
  type InferMotionValueType<T> = T extends MotionValue<infer R> ? R : never
  type InputValues<I extends MotionValue[]> = {
    [K in keyof I]: InferMotionValueType<I[K]>
  }

  declare type MultiTransformer<I extends MotionValue[], O> = (input: InputValues<[...I]>) => O

  export declare function useTransform<I extends MotionValue[], O>(
    input: [...I],
    transformer: MultiTransformer<[...I], O>
  ): MotionValue<O>
}
```

It uses variadic tuples and requires TS version >=4.0. 

PS Feel free to make PR from this script.

**Comment by MikeMajara** [comment]
Until this is merged, specifying `any` type as @huntercaron mentioned gets the build going.
